snapcraft_legacy/storeapi/v2/validation_sets.py

```diff
@@ -14,161 +14,135 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-from typing import Any, Dict, List, Optional
+from typing import Any, Annotated, Dict, Literal, TYPE_CHECKING
+import numbers
+import pydantic
 
-import jsonschema
+from craft_application import models
 
-from ._api_schema import BUILD_ASSERTION_JSONSCHEMA
 
+if TYPE_CHECKING:
+    SnapName = str
+    SnapId = str
+else:
+    SnapName = pydantic.constr(max_length=40)
+    SnapId = pydantic.constr(max_length=40)
 
-class Snap:
+
+def cast_dict_scalars_to_strings(data: dict) -> dict[str, Any]:
+    """Cast all scalars in a dictionary to strings.
+
+    Supported scalar types are str, bool, and numbers.
+    """
+    return {_to_string(key): _to_string(value) for key, value in data.items()}
+
+
+def _to_string(
+    data: dict | list | str | int | float | str | bool | None
+) -> dict[str, Any] | list | str | None:
+    """Recurse through nested dicts and lists and cast scalar values to strings.
+
+    Supported scalar types are str, bool, and numbers.
+    """
+    # start with a string as it is the most common scenario
+    if isinstance(data, str):
+        return data
+
+    if isinstance(data, dict):
+        return {_to_string(key): _to_string(value) for key, value in data.items()}
+
+    if isinstance(data, list):
+        return [_to_string(i) for i in data]
+
+    if isinstance(data, (numbers.Number, bool)):
```

Comment:
If we declare everything in the model as strings we could convert non-scalar types during validation, and let Pydantic's type coercion to handle scalar types (unless we operate on these values and need the original types). Would that work?

Reply:
Discussed offline - the data needs to be an int for some API calls and strings for others, so we will use the correct types in the pydantic model.